main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,381 @@
+  }
+
+  object Expression {
+    def CheckPats(root: TypedAst.Root, tast: TypedAst.Expression)(implicit genSym: GenSym): Validation[TypedAst.Expression, CompilationError] = tast match {
+      case TypedAst.Expression.Match(exp, rules, tpe, loc) =>
+        for  {
+          _rules <- CheckRules(root, exp, rules)
+        } yield TypedAst.Expression.Match(exp, _rules, tpe, loc)
+
+
+      case a => Success(a, Stream.empty)
+    }
+
+    def CheckRules(root: TypedAst.Root, exp: TypedAst.Expression, rules: List[TypedAst.MatchRule]): Validation[List[TypedAst.MatchRule], CompilationError] = {
+      FindNonMatchingPat(root, exp, rules.map(r => List(r.pat)), 1) match {
+        case Right(a) => Success(rules, Stream.empty)
+        case Left((c, rule)) => Failure(ExhaustiveMatchError(formatRule(rule), rule.head.loc.source) #:: Stream.empty);
+      }
+    }
+
+    /**
+      * Format a pattern for printing so the user knows what they missed
+      *
+      * @param rule The rule to print
+      * @return The formatted string
+      */
+    def formatRule(rule: List[TypedAst.Pattern]): String = {
+      def formatPattern(pat: TypedAst.Pattern, acc: String):String = pat match {
+        case Pattern.Int8(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Int16(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Int32(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Int64(lit, _) => acc ++ ", " + lit.toString
+        case Pattern.Float32(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Float64(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Char(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.BigInt(lit, _) => acc ++ ", " ++ lit.toString
+        case Pattern.Wild(_, _) => acc ++ ", _"
+        case Pattern.Var(_, _, _) => acc ++ ", _"
+        case Pattern.Tuple(pats, _ ,_ ) => acc ++ ", (" ++ formatRule(pats) ++ ")"
+        case Pattern.Tag(sym, tag, pt, _, _) => acc ++ ", " ++ sym.name ++ "/" ++ tag ++ formatRule(List(pt))
+      }
+      rule.foldRight("")(formatPattern)
+    }
+
+    /**
+      * Given a list of patterns, computes a pattern vector of size n such
+      * that p doesn't match any rule in rules
+      *
+      * @param  root The AST root of the expression
+      * @param  exp The expression the patterns match against
+      * @param rules The rules to check for exhaustion
+      * @param n The size of resulting pattern vector
+      * @return If no such pattern exists, returns Right(the input), else returns Left(a matching pattern)
+      */
+    def FindNonMatchingPat(root: TypedAst.Root, exp: TypedAst.Expression, rules: List[List[TypedAst.Pattern]], n: Integer): Either[(EnumSym, List[Pattern]), List[List[Pattern]]] = {
+      if (n==0) {
+        if (rules.isEmpty) {
+          return Left((new EnumSym(List.empty[String], "", null), List.empty[Pattern]))
```

Comment:
Woah.. You are not supposed to construct enum symbols... ever. The Namer is the only phase that constructs symbols.

Probably you either need to carry the right enum symbol or to use a custom ADT to represent the data you have.